### ISSUE TITLE
`torch.compile` produce wrong backward gradient result in `pad` when `mode=reflect`

### ISSUE THREAD

**Issue by FatPigeorz** [pytorch/pytorch#100793]
### 🐛 Describe the bug

torch.compile leads to wrong results in the following program. Please see the comments for the results of some variants.

~~~ python
import torch
import torch.nn as nn
import copy

torch.manual_seed(0)
class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.v9 = nn.Parameter(torch.ones(size=[7, 37, 7, 2], dtype=torch.float32))


    def forward(self):
        v4 = torch.nn.functional.pad(self.v9, pad=[-1, 0, 6, 6], mode='reflect')
        # ✅ with mode = 'constant', 'replicate' or 'circular'
        # ❌ with mode = 'reflect'
        return v4
    
model = Model()
model_tmp = copy.deepcopy(model)
model_tmp.train()
compiled = torch.compile(model_tmp, fullgraph=True)
params = {k: v for k, v in compiled.named_parameters()}
ret_compiled = compiled()
grad_compiled = torch.autograd.grad(
                    outputs=ret_compiled,
                    inputs=params.values(),
                    grad_outputs=torch.ones_like(ret_compiled),
                    retain_graph=True,
                    allow_unused=True,
                )
print(grad_compiled)

model.train()
params = {k: v for k, v in model.named_parameters()}
ret_eager = model()
grad_eager = torch.autograd.grad(
                    outputs=ret_eager,
                    inputs=params.values(),
                    grad_outputs=torch.ones_like(ret_eager),
                    retain_graph=True,
                    allow_unused=True,
                )
print(grad_eager)


torch.testing.assert_close(ret_compiled, ret_eager)
# ✅ direct output is correct
torch.testing.assert_close(grad_compiled, grad_eager, rtol=1e-2, atol=1e-3)
# ❌ the gradient calculated with backward ad is wrong

# Mismatched elements: 1813 / 3626 (50.0%)
# Greatest absolute difference: 3.0 at index (0, 0, 1, 0) (up to 0.001 allowed)
# Greatest relative difference: inf at index (0, 0, 0, 0) (up to 0.01 allowed)
# grad_compiled =  
# (tensor([[[[2., 2.],
#          [3., 3.],
#          [3., 3.],
#           ...,))
#
# grad_eager =
# (tensor([[[[0., 2.],
#          [0., 3.],
#          [0., 3.],
#          ...,)


~~~

### Versions

<details><summary>Environment. <i>[Click to expand].</i></summary>
<div>
Collecting environment information...
PyTorch version: 2.1.0.dev20230505+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: 10.0.0-4ubuntu1
CMake version: version 3.26.3
Libc version: glibc-2.31

Python version: 3.10.10 (main, Mar 21 2023, 18:45:11) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-69-generic-x86_64-with-glibc2.31
Is CUDA available: False
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 2080 Ti
Nvidia driver version: 530.30.02
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] torch==2.1.0.dev20230505+cpu
[pip3] triton==2.0.0
[conda] numpy                     1.24.3                   pypi_0    pypi
[conda] torch                     2.1.0.dev20230505+cpu          pypi_0    pypi
[conda] triton                    2.0.0                    pypi_0    pypi
</div>
</details>

cc @ezyang @gchanan @zou3519 @soumith @msaroufim @wconstab @ngimel @bdhirsh @anijain2305

**Comment by soulitzer** [comment]
@ngimel  maybe this is reproducible on CUDA as well

**Comment by ezyang** [comment]
It repros on cuda, but not with aot_eager_decomp_partition, so it is some sort of generic inductor model.
